.github/workflows/branches-and-prs.yml

```diff
@@ -45,21 +45,35 @@ jobs:
           key: test-${{ runner.os }}-variant-${{ matrix.variant }}-gradle-${{ hashFiles('**/*.gradle', '**/*.gradle.kts', '**/gradle.properties', 'gradle/**', 'buildSrc/src/main/**') }}
           restore-keys: |
             test-${{ runner.os }}-variant-${{ matrix.variant }}-
-      - name: 'Set up JDK'
-        uses: actions/setup-java@v1
+      - name: 'Set up JDK ${{ matrix.java }}'
+        uses: actions/setup-java@v2
+        if: matrix.java != 8
         with:
+          distribution: 'adopt'
           java-version: ${{ matrix.java }}
+      - name: 'Prepare JDK${{ matrix.java }} env var'
+        shell: bash
+        run: echo "JDK${{ matrix.java }}=$JAVA_HOME" >> $GITHUB_ENV
+      - name: 'Set up JDK 8'
+        uses: actions/setup-java@v2
+        with:
+          java-version: 8
+          distribution: 'adopt'
+      - name: Prepare JDK8 env var
+        shell: bash
+        run: echo "JDK8=$JAVA_HOME" >> $GITHUB_ENV
```

Comment:
I haven't played with Java 16 on GH Actions before, so probably I miss something, but why you have separate blocks for Java 8 and others?

Reply:
WDYM? With toolchains we can now properly build the `main` sources with JDK 8, while executing the tests with whatever JDK we want, which matches the real world much more closely. To make the JDKs properly discoverable we have to put them in env vars so gradle can pick them up.

Reply:
Ok, I missed that you setup two JDK versions 8 and X which was problematic without toolchains. That might make sense as is closer to reality (having released Spock built with JDK 8).